zng/cc.rs

```diff
@@ -0,0 +1,441 @@
+                for def in &[
+                    "X86_AVX512",
+                    "X86_MASK_INTRIN",
+                    "X86_AVX512VNNI",
+                    "X86_VPCLMULQDQ_CRC",
+                ] {
+                    cfg.define(def, None);
+                }
+
+                cfg.append(
+                    Some("arch/x86"),
+                    &["adler32_avx512", "adler32_avx512_vnni", "crc32_vpclmulqdq"],
+                );
+
+                if cfg.is_msvc {
+                    cfg.flag("/arch:AVX512");
+                } else {
+                    // The zlib-ng cmake scripts to check target features claim that GCC doesn't
+                    // generate good code unless mtune is set, not sure if this is still the
+                    // case, but we faithfully replicate it just in case
+                    for flag in &[
+                        "-mavx512f",
+                        "-mavx512dq",
+                        "-mavx512bw",
+                        "-mavx512vl",
+                        "-mavx512vnni",
+                        "-mvpclmulqdq",
+                        "-mtune=cascadelake",
+                    ] {
+                        cfg.flag(flag);
+                    }
+                }
+            }
+
+            // Misc
+            cfg.define("X86_PCLMULQDQ_CRC", None);
+            cfg.append(Some("arch/x86"), &["crc32_pclmulqdq"]);
+            cfg.mflag("-mpclmul", None);
+            cfg.mflag("-mxsave", None);
+        }
+        "aarch64" | "arm" => {
+            let is_aarch64 = arch == "aarch64";
+
+            cfg.define("ARM_FEATURES", None);
+            cfg.file("src/zlib-ng/arch/arm/arm_features.c");
+
+            let target_os = env::var("CARGO_CFG_TARGET_OS").unwrap();
+
+            // Support runtime detection on linux/android
+            if matches!(target_os.as_str(), "linux" | "android") {
+                for def in &["HAVE_SYS_AUXV_H", "ARM_AUXV_HAS_CRC32"] {
+                    cfg.define(def, None);
+                }
+
+                if !is_aarch64 {
+                    cfg.define("ARM_AUXV_HAS_NEON", None);
+                }
+            }
+
+            // According to the cmake macro, MSVC is missing the crc32 intrinsic
+            // for arm, don't know if that is still true though
+            if !cfg.is_msvc || is_aarch64 {
+                cfg.define("ARM_ACLE", None).define("HAVE_ARM_ACLE_H", None);
+                cfg.append(Some("arch/arm"), &["crc32_acle", "insert_string_acle"]);
+                // When targeting aarch64 we already need to specify +simd, so
+                // we do that once later in this block
+                if !is_aarch64 {
+                    cfg.mflag("-march=armv8-a+crc", None);
+                    cfg.define("ARM_ASM_HWCAP", None);
+                }
+            }
+
+            // neon
+            cfg.define("ARM_NEON", None);
+
+            // TODO: These intrinsics were only added in gcc 9.4, which is _relatively_
```

Comment:
Possible, sure, it was what it was doing before, but others were concerned about complexity so I removed it. I think it makes sense to have such compile checks in a separate project so that people can add them collaboratively rather than being siloed in a single project, making it easier for other projects to remove cmake as well.

Reply:
I see - then I think it's fair to update the comment to reflect that this compile-time feature check should come from a separate crate so it's not done here. I also think it makes sense to do that - something I try to avoid is TODO's that probably stay forever - so maybe that should even be a `NOTE` explaining why certain flags aren't set here.